aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -504,30 +504,48 @@ Tensor trace_cpu(const Tensor& self) {
   Tensor result;
   ScalarType dtype = get_dtype(result, self, c10::nullopt, true);
   result = at::empty({}, self.options().dtype(dtype));
-  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "trace", [&] {
-    using accscalar_t = at::acc_type<scalar_t, false>;
-    accscalar_t sum = 0;
-    const auto* t_data = self.data_ptr<scalar_t>();
+  if (result.scalar_type() == at::kLong) {
```

Comment:
Hm, is this test right? What about other integral types besides Long?

Reply:
Surprisingly this is kinda the right test.

```
if (dtype == at::kLong) {
  ...
}
```

might be clearer. Add a comment here explaining that all integer types (including bool) will return kLong from `get_dtype()`, and that this branch handles integer types.

Reply:
Why is the new branch needed, however? It seems like, as in the previous kernel, the only different is at the assignment to result at the end?

Reply:
yeah it is a bit confusing with no comment, so I added one for clarity.

Reply:
@mruberry this is why https://github.com/pytorch/pytorch/pull/50380/files#r555248150